File: api/v1/schemas/data_center.py
# -*- coding: utf-8 -*-
"""Data center response schemas."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class DataCenterSourceItem(BaseModel):
    name: str = Field(..., description="数据来源名称")
    count: int = Field(..., description="记录数量")


class DataCenterDatabaseInfo(BaseModel):
    path: Optional[str] = Field(None, description="本地数据库路径")
    exists: bool = Field(False, description="文件是否存在")
    size_bytes: int = Field(0, description="文件大小，字节")
    size_label: str = Field("0 B", description="可读文件大小")


class DataCenterFileInfo(BaseModel):
    key: str
    label: str
    path: str
    exists: bool
    file_count: int
    size_bytes: int
    size_label: str


class DataCenterRecommendation(BaseModel):
    level: str = Field(..., description="info/warning/success")
    title: str
    description: str


class DataCenterOverviewResponse(BaseModel):
    generated_at: str
    database: DataCenterDatabaseInfo
    market_data: Dict[str, Any]
    analysis: Dict[str, Any]
    backtests: Dict[str, Any]
    portfolio: Dict[str, Any]
    news: Dict[str, Any]
    fundamentals: Dict[str, Any]
    files: List[DataCenterFileInfo]
    maintenance: Dict[str, Any] = Field(default_factory=dict)
    ai_routing: Dict[str, Any] = Field(default_factory=dict)
    recommendations: List[DataCenterRecommendation]
    warnings: List[str] = Field(default_factory=list)


class DataCenterPortfolioBacktestRequest(BaseModel):
    force: bool = Field(False, description="是否强制重算已存在的回测结果")
    eval_window_days: Optional[int] = Field(None, ge=1, le=120, description="评估窗口天数")
    min_age_days: Optional[int] = Field(None, ge=0, le=3650, description="只回测至少多少天前的分析")
    limit_per_symbol: int = Field(50, ge=1, le=500, description="每只持仓最多处理多少条历史分析")


class DataCenterMarketDataRefreshRequest(BaseModel):
    force: bool = Field(False, description="是否强制按完整回看窗口刷新")
    lookback_days: Optional[int] = Field(None, ge=30, le=3650, description="首次补齐回看天数")
    refresh_overlap_days: Optional[int] = Field(None, ge=0, le=365, description="已有数据时回刷最近几天")
    max_symbols: Optional[int] = Field(None, ge=1, le=1000, description="本次最多处理多少只标的")


class DataCenterMarketDataRefreshResponse(BaseModel):
    generated_at: str
    status: str
    lookback_days: int
    refresh_overlap_days: int
    max_symbols: int
    force: bool
    end_date: str
    totals: Dict[str, int]
    items: List[Dict[str, Any]]
    ledger_path: Optional[str] = None


class DataCenterPortfolioReviewRequest(BaseModel):
    run_backtests: bool = Field(True, description="生成复盘前是否先执行持仓回测")
    send_notification: Optional[bool] = Field(None, description="是否发送通知，空值使用系统配置")


class DataCenterPortfolioReviewResponse(BaseModel):
    generated_at: str
    report_date: str
    status: str
    json_path: Optional[str] = None
    markdown_path: Optional[str] = None
    notification_sent: bool = False
    portfolio: Dict[str, Any] = Field(default_factory=dict)
    market_data: Dict[str, Any] = Field(default_factory=dict)
    backtest: Optional[Dict[str, Any]] = None
    radar: Dict[str, Any] = Field(default_factory=dict)
    ai_routing: Dict[str, Any] = Field(default_factory=dict)


class DataCenterCleanupRequest(BaseModel):
    dry_run: bool = Field(False, description="只预览不删除")


class DataCenterCleanupResponse(BaseModel):
    generated_at: str
    status: str
    dry_run: bool
    log_retention_days: int
    cache_retention_days: int
    totals: Dict[str, int]
    items: List[Dict[str, Any]]
    ledger_path: Optional[str] = None


class DataCenterPortfolioBacktestItem(BaseModel):
    code: str
    status: str
    message: str
    candidate_count: int
    processed: int
    saved: int
    completed: int
    insufficient: int
    errors: int
    summary: Optional[Dict[str, Any]] = None


class DataCenterPortfolioBacktestResponse(BaseModel):
    generated_at: str
    holding_count: int
    processed_symbols: int
    totals: Dict[str, int]
    items: List[DataCenterPortfolioBacktestItem]


class DataCenterPortfolioRiskRadarItem(BaseModel):
    code: str
    quantity: float
    market_value_base: float
    updated_at: Optional[str] = None
    tone: str
    label: str
    title: str
    message: str
    total_evaluations: int
    completed_count: int
    insufficient_count: int
    win_rate_pct: Optional[float] = None
    avg_simulated_return_pct: Optional[float] = None
    summary: Optional[Dict[str, Any]] = None


class DataCenterPortfolioRiskRadarResponse(BaseModel):
    generated_at: str
    holding_count: int
    items: List[DataCenterPortfolioRiskRadarItem]

File: apps/dsa-web/src/types/dataCenter.ts
export interface DataCenterSourceItem {
  name: string;
  count: number;
}

export interface DataCenterDatabaseInfo {
  path?: string | null;
  exists: boolean;
  sizeBytes: number;
  sizeLabel: string;
}

export interface DataCenterMarketData {
  stockCount?: number;
  barCount?: number;
  firstDate?: string | null;
  latestDate?: string | null;
  dataSources?: DataCenterSourceItem[];
  warehouse?: {
    status?: string;
    path?: string;
    generatedAt?: string | null;
    endDate?: string | null;
    totals?: Record<string, number>;
  };
  warehouseTargets?: {
    targetCount?: number;
    targets?: Array<{
      code: string;
      sources: string[];
      coverage?: {
        barCount?: number;
        firstDate?: string | null;
        latestDate?: string | null;
      };
    }>;
  };
  quality?: {
    status?: string;
    latestDate?: string | null;
    staleCutoffDate?: string | null;
    summary?: {
      fresh?: number;
      stale?: number;
      missing?: number;
      skipped?: number;
    };
    items?: Array<{
      code: string;
      status: 'fresh' | 'stale' | 'missing' | 'skipped' | string;
      label: string;
      message: string;
      barCount: number;
      latestDate?: string | null;
      daysBehind?: number | null;
      sources: string[];
    }>;
  };
}

export interface DataCenterAnalysis {
  reportCount?: number;
  stockCount?: number;
  latestCreatedAt?: string | null;
}

export interface DataCenterBacktests {
  resultCount?: number;
  summaryCount?: number;
  stockCount?: number;
  latestEvaluatedAt?: string | null;
}

export interface DataCenterPortfolio {
  accountCount?: number;
  activeAccountCount?: number;
  positionCount?: number;
  tradeCount?: number;
  snapshotCount?: number;
  latestUpdatedAt?: string | null;
}

export interface DataCenterNews {
  itemCount?: number;
  stockCount?: number;
  latestFetchedAt?: string | null;
}

export interface DataCenterFundamentals {
  snapshotCount?: number;
  stockCount?: number;
  latestCreatedAt?: string | null;
}

export interface DataCenterFileInfo {
  key: string;
  label: string;
  path: string;
  exists: boolean;
  fileCount: number;
  sizeBytes: number;
  sizeLabel: string;
}

export interface DataCenterRecommendation {
  level: 'info' | 'warning' | 'success' | string;
  title: string;
  description: string;
}

export interface DataCenterPortfolioBacktestRequest {
  force?: boolean;
  evalWindowDays?: number;
  minAgeDays?: number;
  limitPerSymbol?: number;
}

export interface DataCenterPortfolioBacktestItem {
  code: string;
  status: 'ok' | 'error' | string;
  message: string;
  candidateCount: number;
  processed: number;
  saved: number;
  completed: number;
  insufficient: number;
  errors: number;
  summary?: Record<string, unknown> | null;
}

export interface DataCenterPortfolioBacktestResponse {
  generatedAt: string;
  holdingCount: number;
  processedSymbols: number;
  totals: {
    candidateCount?: number;
    processed?: number;
    saved?: number;
    completed?: number;
    insufficient?: number;
    errors?: number;
  };
  items: DataCenterPortfolioBacktestItem[];
}

export interface DataCenterMarketDataRefreshRequest {
  force?: boolean;
  lookbackDays?: number;
  refreshOverlapDays?: number;
  maxSymbols?: number;
}

export interface DataCenterMarketDataRefreshResponse {
  generatedAt: string;
  status: string;
  lookbackDays: number;
  refreshOverlapDays: number;
  maxSymbols: number;
  force: boolean;
  endDate: string;
  totals: {
    targetCount?: number;
    processed?: number;
    succeeded?: number;
    failed?: number;
    rowsFetched?: number;
    rowsInserted?: number;
  };
  items: Array<Record<string, unknown>>;
  ledgerPath?: string | null;
}

export interface DataCenterPortfolioReviewRequest {
  runBacktests?: boolean;
  sendNotification?: boolean | null;
}

export interface DataCenterPortfolioReviewResponse {
  generatedAt: string;
  reportDate: string;
  status: string;
  jsonPath?: string | null;
  markdownPath?: string | null;
  notificationSent: boolean;
  portfolio?: Record<string, unknown>;
  marketData?: Record<string, unknown>;
  backtest?: Record<string, unknown> | null;
  radar?: Record<string, unknown>;
  aiRouting?: Record<string, unknown>;
}

export interface DataCenterCleanupRequest {
  dryRun?: boolean;
}

export interface DataCenterCleanupResponse {
  generatedAt: string;
  status: string;
  dryRun: boolean;
  logRetentionDays: number;
  cacheRetentionDays: number;
  totals: {
    scannedMatches?: number;
    deletedCount?: number;
    freedBytes?: number;
    errorCount?: number;
  };
  items: Array<Record<string, unknown>>;
  ledgerPath?: string | null;
}

export interface DataCenterPortfolioRiskRadarItem {
  code: string;
  quantity: number;
  marketValueBase: number;
  updatedAt?: string | null;
  tone: 'strong' | 'watch' | 'weak' | 'empty' | 'error' | string;
  label: string;
  title: string;
  message: string;
  totalEvaluations: number;
  completedCount: number;
  insufficientCount: number;
  winRatePct?: number | null;
  avgSimulatedReturnPct?: number | null;
  summary?: Record<string, unknown> | null;
}

export interface DataCenterPortfolioRiskRadarResponse {
  generatedAt: string;
  holdingCount: number;
  items: DataCenterPortfolioRiskRadarItem[];
}

export interface DataCenterOverview {
  generatedAt: string;
  database: DataCenterDatabaseInfo;
  marketData: DataCenterMarketData;
  analysis: DataCenterAnalysis;
  backtests: DataCenterBacktests;
  portfolio: DataCenterPortfolio;
  news: DataCenterNews;
  fundamentals: DataCenterFundamentals;
  files: DataCenterFileInfo[];
  maintenance?: {
    portfolioDailyReview?: {
      status?: string;
      path?: string;
      markdownPath?: string;
      generatedAt?: string | null;
      reportDate?: string | null;
      holdingCount?: number;
    };
    cleanup?: {
      status?: string;
      path?: string;
      generatedAt?: string | null;
      totals?: Record<string, number>;
    };
  };
  aiRouting?: {
    mode?: string;
    cloudAnalysis?: {
      enabled?: boolean;
      providerCount?: number;
      providers?: string[];
      primaryModel?: string;
      configuredModels?: string[];
      role?: string;
    };
    localWorkstation?: {
      enabled?: boolean;
      role?: string;
      tasks?: string[];
    };
    localModel?: {
      defaultEnabled?: boolean;
      role?: string;
    };
    recommendation?: string;
  };
  recommendations: DataCenterRecommendation[];
  warnings: string[];
}
